backend-api/src/functions/adapters/tests/dynamoDbAdapter.test.ts

```diff
@@ -0,0 +1,150 @@
+import { expect } from "@jest/globals";
+import "dotenv/config";
+import "../../testUtils/matchers";
+import { DynamoDbAdapter } from "../dynamoDbAdapter";
+import { BiometricTokenIssued } from "../../common/session/updateOperations/BiometricTokenIssued/BiometricTokenIssued";
+import {
+  SessionRegistry,
+  UpdateSessionError,
+} from "../../common/session/SessionRegistry";
+import {
+  ConditionalCheckFailedException,
+  DynamoDBClient,
+  UpdateItemCommand,
+} from "@aws-sdk/client-dynamodb";
+import { mockClient } from "aws-sdk-client-mock";
+import { emptySuccess, errorResult, Result } from "../../utils/result";
+import { UpdateSessionOperation } from "../../common/session/updateOperations/UpdateSessionOperation";
+
+const mockDynamoDbClient = mockClient(DynamoDBClient);
+
+let sessionRegistry: SessionRegistry;
+let consoleErrorSpy: jest.SpyInstance;
+let consoleDebugSpy: jest.SpyInstance;
+
+describe("DynamoDbAdapter", () => {
+  beforeEach(() => {
+    sessionRegistry = new DynamoDbAdapter("mock_table_name");
+    consoleErrorSpy = jest.spyOn(console, "error");
+    consoleDebugSpy = jest.spyOn(console, "debug");
+  });
+  describe("updateSession", () => {
+    let result: Result<void, UpdateSessionError>;
+
+    const updateOperation: UpdateSessionOperation = new BiometricTokenIssued(
+      "NFC_PASSPORT",
+      "mock_opaque_id",
+    );
+
+    describe("On every attempt", () => {
+      beforeEach(async () => {
+        await sessionRegistry.updateSession("mock_session_id", updateOperation);
+      });
+
+      it("Logs the attempt", () => {
+        expect(consoleDebugSpy).toHaveBeenCalledWithLogFields({
+          messageCode: "MOBILE_ASYNC_UPDATE_SESSION_ATTEMPT",
+          data: {
+            updateExpression: updateOperation.getDynamoDbUpdateExpression(),
+            conditionExpression:
+              updateOperation.getDynamoDbConditionExpression(),
+          },
+        });
+      });
+    });
+
+    describe("When a conditional check fails", () => {
+      beforeEach(async () => {
+        mockDynamoDbClient.on(UpdateItemCommand).rejects(
+          new ConditionalCheckFailedException({
+            $metadata: {},
+            message: "Conditional check failed",
+          }),
+        );
+        result = await sessionRegistry.updateSession(
+          "mock_session_id",
+          updateOperation,
+        );
+      });
+
+      it("Logs the failure", () => {
+        expect(consoleErrorSpy).toHaveBeenCalledWithLogFields({
+          messageCode: "MOBILE_ASYNC_UPDATE_SESSION_CONDITIONAL_CHECK_FAILURE",
+          error: "Conditional check failed",
+          data: {
+            updateExpression: updateOperation.getDynamoDbUpdateExpression(),
+            conditionExpression:
+              updateOperation.getDynamoDbConditionExpression(),
+          },
+        });
+      });
+
+      it("Returns failure with conditional check failure error", () => {
```

Comment:
Tiny nit, it doesn't actually return the failure here

```suggestion
      it("Returns errorResult with conditional check failure error", () => {
```

Reply:
This would go for the other "Returns failure..." blocks in this file. Feel free to dismiss

Reply:
That's just me being set in my ways with MPT nomenclature. I may keep this as is, though, because even though the helper function is still called `errorResult` (one day I'll look to get that changed), the type is (now) called Failure/FailureWithValue/EmptyFailure